include/graaflib/algorithm/shortest_path.tpp

```diff
@@ -61,6 +61,80 @@ std::optional<GraphPath<int>> get_unweighted_shortest_path(
   }
 }
 
+template <typename V, typename E, graph_spec S>
+std::optional<GraphPath<E>> get_weighted_shortest_path(
+    const graph<V, E, S>& graph, vertex_id_t start_vertex,
+    vertex_id_t end_vertex) {
+  using weight_t = typename graph<V, E, S>::edge_t;
```

Comment:
I do not get the `no match for ‘operator+’ (operand types are ‘weight_t’ {aka ‘std::shared_ptr<graaf::primitive_numeric_adapter<int> >’} and ‘int’)` error locally, but I suspect this only happens when this new function is called in a test.

I think this could also be related to the current line. `typename(graph)::edge_t` only yields us a `shared_ptr`, see `graph.h` line 38:

```c++
using edge_t =
      std::conditional_t<detail::is_primitive_numeric_v<EDGE_T>,
                         std::shared_ptr<primitive_numeric_adapter<EDGE_T>>,
                         std::shared_ptr<EDGE_T>>;
```

This shared pointer either wraps `primitive_numeric_adapter<EDGE_T>` (if `EDGE_T` is a primitive numeric) or `EDGE_T` otherwise. To get this wrapped type out of the shared pointer we can do `typename decltype(graph)::edge_t::element_type`.

Both `primitive_numeric_adapter<EDGE_T>` as well as `EDGE_T` will derive from `weighted_edge` if we are dealing with a weighted graph. We can get a weight type out of the weighted edge with the trait `weight_t`.

Combining this all:

```c++
using edge_ptr_t = typename decltype(graph)::edge_t;
using wrapped_t = typename edge_ptr_t::element_type;
using weight_t = typename wrapped_t::weight_t;

// Or all combined in one line
using weight_t = typename decltype(graph)::edge_t::element_type::weight_t;
```

Reply:
With a minimal test, this proposal fails for me with `error: ‘const graaf::graph<int, graaf::algorithm::my_weighted_edge<int>, graaf::graph_spec::UNDIRECTED>&’ is not a class, struct, or union type`. 
After some googling, adding `std::remove_reference_t` like
```cpp
using edge_ptr_t = typename std::remove_reference_t<decltype(graph)>::edge_t;
```
seems to prevent an error on this exact line. I'm not aware if this change might introduce unintended side effects.
Please note that with the tests provided, the project still doesn't build successfully due to
`from ‘graaf::algorithm::GraphPath<int>’ to ‘std::optional<graaf::algorithm::GraphPath<graaf::algorithm::my_weighted_edge<int> > >` that probably relates to the shared pointer architecture as well. The test is basically your setup for testing weighted edges reduced to integer types.
I pushed my working state in d78e821 for your reference. Help is highly appreciated! Feel free to push to this fork directly if you see an easy fix.